### PR TITLE
[DQMOffline/Trigger] added "requireValidHLTPaths" flag to plugins using GenericTriggerEventFlag

### PR DESCRIPTION
**Pull request by missirol**
#### PR description:

This PR propagates the check on valid HLT paths (originally introduced in #27772) to more plugins of the HLT Offline DQM; specifically, the feature is added to the plugins that make use of the `GenericTriggerEventFlag` class.

The changes here are limited to the package `DQMOffline/Trigger`, so they only apply to the (HLT) DQM step (and not the Validation step).

While implementing this, it became apparent that several plugins contained exact copies of the same methods (e.g. `setMETitle`); luckily, many of these basic methods were already part of the class `TriggerDQMBase`, so to minimize redundant code, many plugins have been updated to inherit from `TriggerDQMBase` (as was/is done for the most recent plugins in this same package).

The changes, in terms of lines of code, turned out to be fairly extensive, but they are in principle straightforward; aside from the addition of the "valid HLT paths" functionality and the inheritance from `TriggerDQMBase`, some polishing of the code was done while going through the different plugins (removal of unnecessary `#include`s, etc), but without changes to functionalities nor outputs.

The idea of this PR would be to first run the comparisons to verify that there are no unexpected differences, and then change the default of `requireValidHLTPaths` from `False` to `True` for all possible plugins (this could be done here, or in a separate PR; my preference would be to do it here after the first set of comparisons is understood), to check how many (empty) DQM histograms are removed after doing this (to be clear, these would be DQM outputs that are currently produced in the different workflows, but inevitably empty because associated to HLT paths not present in the HLT Menu in the input). I ran one check on a 2018 MC file and, there, switching to `requireValidHLTPaths = True` resulted in ~3500 less (HLT-)DQM histograms.

Other small changes include moving `TriggerDQMBase.*` out of `plugins/` (since it's not a plugin), and removing `plugins/TopMonitor.h` (moving the relevant code to `plugins/TopMonitor.cc`).

There is one change expected in the comparisons (even with `requireValidHLTPaths = False`): after inheriting from `TriggerDQMBase`, the plugin `NoBPTXMonitor` will produce a few more histograms under `EXO/NoBPTX/*`, because the booking routines in that plugin were slightly different from those it now inherits from `TriggerDQMBase`.

#### PR validation:

Successfully ran a few standard workflows (for example, `runTheMatrix.py -l 11634.0`).

#### if this PR is a backport please specify the original PR:

Not a backport, and no backport of this is planned.

### CODE REVIEW

File: DQMOffline/Trigger/plugins/TopMonitor.h [hunk 1]
```diff
@@ -1,250 +0,0 @@
-#ifndef DQMOffline_Trigger_TopMonitor_h
```
Comment (jfernan2) [inline DQMOffline/Trigger/plugins/TopMonitor.h]: @missirol  I don't understand why in this case you move everything from .h to .cc file while in the other plugins you don't

Comment (missirol) [reply DQMOffline/Trigger/plugins/TopMonitor.h]: it's a fair point, in principle it should be done (i had done it for `TopMonitor` extemporaneously, and this got included in this PR); for the other plugins, i was trying to stick to the changes i originally planned (the ones described in the PR)

Comment (jfernan2) [reply DQMOffline/Trigger/plugins/TopMonitor.h]: @missirol I don't have a preference so I let code experts to evaluate and decide but I believe there should be coherence across the other plugins

**Comment by jfernan2** [comment]
please test
Let's see first if there are unexpected differences

**Comment by missirol** [comment]
@jfernan2 , as far as I can see, at the moment the only difference in the comparisons is the addition, for certain workflows, of 44 histograms under `EXO/NoBPTX/`, which is expected (as mentioned in the PR description)

**Comment by jfernan2** [comment]
@missirol  yes I see, upon pending the decission from the experts about the .h to .cc code migration, are these 44 new histograms needed? They seem no empty but were not present before.

Another question: do you finally plan to use this PR to change the default of requireValidHLTPaths from False to True as you mention in the description? I am in favor of that instead of making another PR. It would be nice to see how many empty histograms are cleaned.

I am also concerned why HLT is still running and producing empty plots despite the FakeHLTMenu option is being used in general but in Phase2 workflows.

Thank you very much for this PR in any case @missirol

**Comment by missirol** [comment]
hi @jfernan2 

as for the 44 histograms, I doubt they are really needed; let me tweak the common methods in `TriggerDQMBase` to avoid this (maintaining backward compatibility); I can do it in this PR by the end of the week.

As for the flag, yes, that's the plan (i had edited my previous post with a comment on this, but maybe after your reply).

As for the Phase-2 workflows, honestly I don't know exactly what's going on there in terms of HLT DQM; if you want to point me to the relevant info, i can take a look (in the new year, most likely).

**Comment by jfernan2** [comment]
Let me clarify: *FakeHLT DQM and Validation sequences are being run everywhere except in the Phase2 samples where new HLT menus will be tested. This means that HLT plots should be in principle removed by this in PR https://github.com/cms-sw/cmssw/pull/27467  since the GRun HLT is not being used

The relevant code is;
https://cmssdt.cern.ch/lxr/source/DQMOffline/Configuration/python/autoDQM.py#0009
https://cmssdt.cern.ch/lxr/search?%21v=CMSSW_11_1_X_2019-12-16-2300&_filestring=&_string=FakeHLT

**Comment by missirol** [comment]
Hi @jfernan2 ,

I pushed a couple more commits to remove the 44 histograms in excess and move the headers of the plugins involved inside their `plugins/*.cc` implementation, following the [CMS Naming, Coding, And Style Rules](https://cms-sw.github.io/cms_coding_rules.html) (to be clear, this is not all the plugins in the package, I stuck to the ones touched by this PR in this first place, I prefer not to touch the other ones for now).

I think we can re-run the comparisons to verify there are no differences, then switch the `requiredValidHLTPaths` flag to `True` and re-test.

**Comment by jfernan2** [comment]
please test

**Comment by jfernan2** [comment]
thanks @missirol !!

**Comment by missirol** [comment]
Hi @jfernan2 , I see no differences, right? Let me know if I can go ahead with changing the flag to `True`.

**Comment by jfernan2** [comment]
@missirol Yes indeed, please proceed. Thanks a lot!

**Comment by makortel** [comment]
> Other small changes include moving `TriggerDQMBase.*` out of `plugins/` (since it's not a plugin), and removing `plugins/TopMonitor.h` (moving the relevant code to `plugins/TopMonitor.cc`).

Are any of the moved files needed outside of the package? If not, we actually recommend to keep the files under `plugins/` (see 6.8 in http://cms-sw.github.io/cms_coding_rules.html#6--packaging-rules-1)

**Comment by missirol** [comment]
@makortel , true, i missed this, thanks.
to my knowledge, the answer is no, so classes like `TriggerDQMBase` and a few others could go back to `plugins/`; i'll do that then (@jfernan2 please don't trigger the tests just yet)

**Comment by missirol** [comment]
@makortel 's comment should be addressed by 63b45736b70c; if so, from my side, we'd be ready to re-test the changes.

**Comment by jfernan2** [comment]
please test

**Comment by missirol** [comment]
@jfernan2 
i think i have to look into the O(10) differences in wfs `1000`, `11634` and `12434`; i didn't expect those.

**Comment by jfernan2** [comment]
@missirol  I was looking exactly at that, I believe it is related to the sample type: 1000 is MiniBias2011 while the other are MC TTBar for Run2 and Run3 (11634 and 12434) but if you can find out why it is better. 
I was also hoping for some reduction on Data samples: 136.X, 140.X and 158.X but those are already cleaned by standardDQMFakeHLT , which could have been an option for the MC above too but the way this PR does is probably more elegant.
Nevertheless, thank you very much again for this PR, many histograms are being cleaned by this change and most of them empty (but some denominator histograms for efficiency calculations where numerator is empty).

In a second step cleaning, in the current baseline I see the folllowing empty histos:
wf4.22: HLT/EGM, HLT/GeneralHLTOffline/Hcal*   HLT/Muon  HLT/JME   HLT/TAU
wf4.53: HLT/EventInfo/muonQuality   HLT/Lumi*
wf7.3: HLT/JME    HLT/EGM   HLT/EventInfo/muonQuality
wf8.0: HLT/JME    HLT/EGM   HLT/EventInfo/muonQuality
wf9.0: HLT/EventInfo/muonQuality   HLT/Lumi*  HLT/EGM   
wf25.0: HLT/EventInfo/muonQuality   HLT/Lumi*  HLT/EGM   
wf136.X:  HLT/EventInfo/muonQuality   HLT/Lumi*
wf140.X:  HLT/EventInfo/muonQuality   HLT/Lumi*

**Comment by missirol** [comment]
@jfernan2 okay!
i will try to look into the first point by the end of the week.
on the second point: noted, and thanks for the info.

**Comment by missirol** [comment]
Hi @jfernan2 

I looked into the O(10) differences found in wfs `1000`, `11634` and `12434`.

I reproduced them locally with a 2018 MC file
(this should be representantive of what happens in `11634` and `12434`;
we can then check if the update I just committed also solves `1000`)

The differences came up because of the following things:

1. there are two sequences, in `HTMonitor_cff` and `METMonitor_cff`, each making use of one plugin (`HTMonitor.cc` and `METMonitor.cc`, respectively);

2. for the folders where the differences came up,
   there are two modules (one in `HTMonitor_cff` and one in `METMonitor_cff`) that write to the same output folder;
   the two modules are mostly redudant in the `cc` implementation, but configured differently in the `cff`s;
   as a consequence, there are inconsistent histograms in each of the folders in question.

3. on top of that, the plugins are similar enough that, in some cases, they book the same histogram (same TH1 name, etc);
   for these cases, the modules in `HTMonitor_cff` overwrite some of the histograms created by `METMonitor_cff`;

4. after switching `requireValidHLTPaths` to `True`, all the `HTMonitor_cff` modules are prevented from creating DQM outputs
   if using a recent 2018 HLT Menu, because they are requiring an HLT path that is not in that menu (`IsoTkMu27`, required here);
   as a consequence, some of the DQM outputs created by `METMonitor_cff` are not overwritten, and this leads to the differences.

For this PR, I would suggest the following:

1. apply the minimal bugfixes to the `HTMonitor_cff` sequence [*]
   to have it working on the 2018 Menu, like the vast majority of the code in this package
   (these fixes might lead to some small differences in the comparisons);

2. leave dedicated changes to these plugins (for example, removing redundant ones) to a future PR
   (tentatively in January, when we should receive the groups' feedback on the actual content of the various folders).

[*] I spotted names of outdated paths in a couple of other modules, and I fixed those as well (see acadddc9e37a).

**Comment by missirol** [comment]
I rebased to solve merge-conflicts with #28651 for one of the DQM monitors touched by this PR

**Comment by jfernan2** [comment]
please test

**Comment by missirol** [comment]
@jfernan2

i think the remaining differences in wf `1000` are due to what i wrote [here](https://github.com/cms-sw/cmssw/pull/28627#issuecomment-568188631); if I checked well, `IsoMu27` is not in that 2011 menu, so the `HTMonitor` module is skipped by the PR (and the histograms in question are then those created by `METMonitor`).

Other than that, I think the rest looks as expected.

**Comment by jfernan2** [comment]
@missirol only wfs 11634 and 12434 have changed, wf1000 remains the same.
BTW: I am fine with the plan for this PR you suggest

**Comment by jfernan2** [comment]
Our comments crossed eachother: I am fine with cleaning you have done.
From your plan:

1.    apply the minimal bugfixes to the HTMonitor_cff sequence [*]
    to have it working on the 2018 Menu, like the vast majority of the code in this package
    (these fixes might lead to some small differences in the comparisons);

2.    leave dedicated changes to these plugins (for example, removing redundant ones) to a future PR
    (tentatively in January, when we should receive the groups' feedback on the actual content of the various folders).

I understand 1. is already in place, and 2. will be done in January at some point.
Thanks

**Comment by missirol** [comment]
@jfernan2 yes, exactly

(to be clear on 2. : some cleanup can definitely be done anyway in January, but the timeline for a full cleanup of the sequences/plugins will depend on when we receive feedback from the different groups)

**Comment by jfernan2** [comment]
+1

**Comment by missirol** [comment]
Rebased to solve merge conflicts introduced by the merge of #28552.

There were no real conflicting changes, just a few lines where git got confuses in the following files
`DQMOffline/Trigger/plugins/{LepHTMonitor.*,TagAndProbeBtagTriggerMonitor.cc}`,
which are touched by both this PR and #28552.

Comparisons are expected to be the same as the last ones ran for this PR.

**Comment by jfernan2** [comment]
please test

**Comment by jfernan2** [comment]
+1
Changes equivalent to those approved in 2019

**Comment by silviodonato** [comment]
+1

**Pull request merged**